### ISSUE TITLE
Implement Display for Rope

### ISSUE THREAD

**Issue by cmyr** [xi-editor/xi-editor#772]
Just noticed while looking at #771 that ~these aren't~ this isn't implemented.

**Comment by zettaittenani** [comment]
I really want to implement this. How about the code below?
Testing passed.

```diff
diff --git a/rust/rope/src/rope.rs b/rust/rope/src/rope.rs
index 21d6410..3978dd2 100644
--- a/rust/rope/src/rope.rs
+++ b/rust/rope/src/rope.rs
@@ -708,6 +708,16 @@ impl fmt::Debug for Rope {
     }
 }

+impl fmt::Display for Rope {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if f.alternate() {
+            write!(f, "{}", String::from(self))
+        } else {
+            write!(f, "Rope({})", String::from(self))
+        }
+    }
+}
+
 impl Add<Rope> for Rope {
     type Output = Rope;
     fn add(self, rhs: Rope) -> Rope {
@@ -1174,7 +1184,7 @@ mod tests {

         let rope_with_emoji = Rope::from("hi\ni'm\n😀 four\nlines");
         let utf16_units = rope_with_emoji.measure::<Utf16CodeUnitsMetric>();
-
+
         assert_eq!(utf16_units, 20);

         // position after 'f' in four
@@ -1193,4 +1203,14 @@ mod tests {
         let utf8_offset = rope_with_emoji.convert_metrics::<Utf16CodeUnitsMetric, BaseMetric>(utf16_units);
         assert_eq!(utf8_offset, 19);
     }
+
+    #[test]
+    fn display() {
+        let rope = Rope::from("hi\ni'm\nfour\nlines");
+        let display_fmt_alt = "hi\ni'm\nfour\nlines";
+        assert_eq!(&format!("{:#}", rope), display_fmt_alt);
+
+        let display_fmt = "Rope(hi\ni'm\nfour\nlines)";
+        assert_eq!(&format!("{}", rope), display_fmt);
+    }
 }
```

**Comment by cmyr** [comment]
@zettaittenani Okay, I think that code looks good!

If you wanted to get fancy, here are some extra ideas:

- instead of `String::from(..)`, you could iterate through the contents of the rope with `Rope::iter_chunks`, and then write these out. This would save having to reallocate the whole contents.

- alt display format: I hadn't thought about this, but now that you've added it, I had an idea: it would be cool if we could print the contents but replace the whitespace with visible characters? e.g. space, tab, `\r` and `\n` could be replaced with `∙`, `⇥`, `␍`, and `␊`? This is something I've needed in the past while debugging, so might be nice to have it built in.

So: would be happy to accept your current patch, and if you'd like to try either of these ideas you're welcome to. Thanks! 🙇

**Comment by jansol** [comment]
@zettaittenani it's been over a month, can we have a status update?

**Comment by zettaittenani** [comment]
@jansol Sorry for late, I will send a patch for the moment.

**Comment by sjoshid** [comment]
@cmyr,
I have tried to implement this. Haven't done anything fancy like ∙, ⇥, ␍, and ␊ that you wanted. Let me know if you do. 
Also I would like to mention I didnt mean to discredit @zettaittenani efforts to fix this. Im a Rust beginner and just wanted to know if my understanding of what you want is correct.

**Comment by cmyr** [comment]
fixed in #830
